### ISSUE TITLE
[Feature] If DB connection is lost, retry to connect to it when it is back up

### ISSUE THREAD

**Issue by rdehuyss** [jobrunr/jobrunr#82]
**Is your feature request related to a problem? Please describe.**
If the DB goes down, JobRunr automatically shuts down. 

**Describe the solution you'd like**
It would be nice that if the DB becomes available again, JobRunr would automatically restart again.

**Describe alternatives you've considered**
Stop/Start manually?

**Comment by jonevn** [comment]
How about some sort of backoff? Say that the connection goes down, then first try again in a few seconds and for every fail, then add a decent amount of time before trying again. Finally it could test once every ~5 minutes forever.

Would also be nice to have an indicator on the dashboard for DB-connection status. And the status of the DB-connection should off course be saved in the DB =)

**Comment by junhuhdev** [comment]
@rdehuyss Question is there a reason that the backgroundJobServer needs to be shutdown? 

Can't we instead then store future jobs in an inmemory storage until the db connection is back up?

**Comment by rdehuyss** [comment]
@jonevn - I like the idea of the backoff. Time is scarce for the moment, so I don't know when I will implement this. It may also become a feature of the Pro version as normally, a database is typically up & running most of the time.
@junhuhdev - The reason for the shutdown is that otherwise your logs will be filled with a lot of errors of JobRunr trying to process jobs. As it can't update them though, a lot does not work:
- work distribution
- updates to jobs
- ...

Storing them in-memory is also not an option for me as you typically want to be sure that your jobs are processed only once (think email sending). As the database is used to keep track of what happened, this then won't work anymore.

**Comment by junhuhdev** [comment]
@rdehuyss Thank you for the explanation, I agree with most of the points. Just of general curiosity do you know if hangfire has solved it and if so in what way?

**Comment by junhuhdev** [comment]
@rdehuyss One more question even though I start the backgroundJobServer manually, and enqueue a new job, I see the new one in the dashboard. But it doesn't get executed anymore.

**Comment by junhuhdev** [comment]
this is the example code ive been testing with. 

Issue with doubled count of workers seems to happen when I set the delay to a smaller amount such as 1000 ms for example. 

But if I increase it the issue seems to disappear. 

```java
	@Scheduled(fixedDelay = 1000 * 60)
	public void aliveCheck() {
		log.info("Alive check...");
		var isRunning = backgroundJobServer.isRunning();
		if (!isRunning) {
			log.info("Starting backgroundJobServer again...");
			backgroundJobServer.start();
		}
	}
```

**Comment by rdehuyss** [comment]
Hmm, I missed your feedback. I'll look into the reason why the worker count doubles if time is less than 1sec.

**Comment by rdehuyss** [comment]
This has been resolved in JobRunr Pro v4.0.0. See the [blog post](https://www.jobrunr.io/en/blog/2021-09-08-jobrunr-v4.0.0/) for more info.

**Comment by Arvindkomal** [comment]
@junhuhdev i tried your approach. it's not working. actually job server re started but not triggering the jobs.
`@Scheduled(cron = "0/180 * * ? * *")
@SchedulerLock(
        name = "job_lock",
        lockAtLeastFor = "30s",
        lockAtMostFor = "40s"
)
@Trace(dispatcher=true)
public void restartJobs(){
    log.error("in Job Runner");
    if(JobRunr.getBackgroundJobServer() != null && !JobRunr.getBackgroundJobServer().isRunning()) {
        JobRunr.getBackgroundJobServer().start();
    }`

**Comment by rdehuyss** [comment]
I do not recommend this approach - your disks will be flooded with logs.

**Comment by Arvindkomal** [comment]
@rdehuyss ok. any idea how to handle db connection lost issue ?

**Comment by rdehuyss** [comment]
See also https://github.com/jobrunr/jobrunr/discussions/334
